### ISSUE TITLE
sparse - sparse matrix multiplication partially missing

### ISSUE THREAD

**Issue by albertomercurio** [JuliaGPU/CUDA.jl#1599]
Hello,

I'm trying to do a matrix multiplication when both matrices are of the sparse CSR type. For example

```julia
using CUDA
CUDA.allowscalar(false)
using SparseArrays

A = sprand(ComplexF32, 1000, 1000, 0.001)
A_gpu = CUSPARSE.CuSparseMatrixCSR(A)
A_gpu * A_gpu
```

returns the error

```julia
ERROR: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore are only permitted from the REPL for prototyping purposes.
If you did intend to index this array, annotate the caller with @allowscalar.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] assertscalar(op::String)
    @ GPUArraysCore ~/.julia/packages/GPUArraysCore/lojQM/src/GPUArraysCore.jl:87
  [3] getindex(xs::CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, I::Int64)
    @ GPUArrays ~/.julia/packages/GPUArrays/fqD8z/src/host/indexing.jl:9
  [4] getindex(A::CUDA.CUSPARSE.CuSparseMatrixCSR{ComplexF32, Int32}, i0::Int64, i1::Int64)
    @ CUDA.CUSPARSE ~/.julia/packages/CUDA/DfvRa/lib/cusparse/array.jl:321
  [5] _generic_matmatmul!(C::Matrix{ComplexF32}, tA::Char, tB::Char, A::CUDA.CUSPARSE.CuSparseMatrixCSR{ComplexF32, Int32}, B::CUDA.CUSPARSE.CuSparseMatrixCSR{ComplexF32, Int32}, _add::LinearAlgebra.MulAddMul{true, true, Bool, Bool})
    @ LinearAlgebra ~/julia-1.8.1/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:876
  [6] generic_matmatmul!(C::Matrix{ComplexF32}, tA::Char, tB::Char, A::CUDA.CUSPARSE.CuSparseMatrixCSR{ComplexF32, Int32}, B::CUDA.CUSPARSE.CuSparseMatrixCSR{ComplexF32, Int32}, _add::LinearAlgebra.MulAddMul{true, true, Bool, Bool})
    @ LinearAlgebra ~/julia-1.8.1/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:844
  [7] mul!
    @ ~/julia-1.8.1/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:303 [inlined]
  [8] mul!
    @ ~/julia-1.8.1/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:276 [inlined]
  [9] *(A::CUDA.CUSPARSE.CuSparseMatrixCSR{ComplexF32, Int32}, B::CUDA.CUSPARSE.CuSparseMatrixCSR{ComplexF32, Int32})
    @ LinearAlgebra ~/julia-1.8.1/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:141
 [10] top-level scope
    @ ~/GitHub/Research/2022/Generation of Phonons/Prova.jl:16
```

A workaround could be definind the function by myself, with something like

```julia
import Base

function Base.:(*)(A::CUSPARSE.CuSparseMatrixCSR, B::CUSPARSE.CuSparseMatrixCSR)
    C = copy(A)
    CUSPARSE.mm!('N', 'N', one(eltype(A)), A, B, zero(eltype(A)), C, 'O')
    C
end

Array(A_gpu * A_gpu) ≈ A * A
```

which give me `true`.

Is there a way to implement directly this feature inside CUDA.jl?

**Comment by maleadt** [comment]
sparse*sparse is relatively new, and that functionality apparently isn't integrated with the LinAlg interfaces yet: https://github.com/JuliaGPU/CUDA.jl/blob/master/lib/cusparse/interfaces.jl. Should be easy to add.

**Comment by albertomercurio** [comment]
Thanks for you quick response.

So in the meantime I have to implement these functions by myself, waiting the official implementation?

**Comment by maleadt** [comment]
Yes. And if you implement then, you could as well create a PR here :-) Note that you'd need to implement `LinearAlgebra.mul!`, not `Base.*`.

**Comment by albertomercurio** [comment]
I tried to implement

```julia
function LinearAlgebra.mul!(Y::CUSPARSE.CuSparseMatrixCSR, A::CUSPARSE.CuSparseMatrixCSR, B::CUSPARSE.CuSparseMatrixCSR)
    CUSPARSE.mm!('N', 'N', one(eltype(A)), A, B, zero(eltype(A)), Y, 'O')
end
```

but it gives me the scalar indexing error. 

This is due to the fact that LinearAlgebra.jl calls something like `mul!(similar(A_gpu, eltype(A_gpu), (size(A_gpu, 1), size(A_gpu, 2))), A_gpu, A_gpu)` but `similar(A_gpu, eltype(A_gpu), (size(A_gpu, 1), size(A_gpu, 2)))` is a CPU matrix and not a CuSparseMatrixCSR, and thus it doesn't call my function, but the generic function of LinearAlgebra.jl


The fatstest workaround I found is to declare

```julia
using SparseArrays

function LinearAlgebra.:(*)(A::CUSPARSE.CuSparseMatrixCSR, B::CUSPARSE.CuSparseMatrixCSR)
    Y = CUSPARSE.CuSparseMatrixCSR(spzeros(eltype(A), size(A, 1), size(B, 2)))
    mul!(Y, A, B)
end
```

**Comment by albertomercurio** [comment]
Closed by #1603
